File: api/src/test/java/org/pmiops/workbench/db/dao/ParticipantCohortAnnotationDaoTest.java
package org.pmiops.workbench.db.dao;

import static com.google.common.truth.Truth.assertThat;

import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.pmiops.workbench.SpringTest;
import org.pmiops.workbench.config.CommonConfig;
import org.pmiops.workbench.db.model.DbCohortAnnotationDefinition;
import org.pmiops.workbench.db.model.DbCohortAnnotationEnumValue;
import org.pmiops.workbench.db.model.DbParticipantCohortAnnotation;
import org.pmiops.workbench.model.AnnotationType;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.context.annotation.Import;
import org.springframework.test.annotation.DirtiesContext;

@Import({CommonConfig.class})
@DataJpaTest
@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
public class ParticipantCohortAnnotationDaoTest extends SpringTest {

  private static long COHORT_ID = 1;
  @Autowired private ParticipantCohortAnnotationDao participantCohortAnnotationDao;
  @Autowired private CohortAnnotationDefinitionDao cohortAnnotationDefinitionDao;
  private DbParticipantCohortAnnotation pca;
  private DbParticipantCohortAnnotation pca1;
  private long cohortReviewId = 3L;
  private long participantId = 4L;

  @BeforeEach
  public void setUp() {
    DbCohortAnnotationDefinition enumAnnotationDefinition =
        new DbCohortAnnotationDefinition()
            .cohortId(COHORT_ID)
            .columnName("enum")
            .annotationTypeEnum(AnnotationType.ENUM);
    DbCohortAnnotationEnumValue enumValue1 =
        new DbCohortAnnotationEnumValue()
            .name("z")
            .order(0)
            .cohortAnnotationDefinition(enumAnnotationDefinition);
    DbCohortAnnotationEnumValue enumValue2 =
        new DbCohortAnnotationEnumValue()
            .name("r")
            .order(1)
            .cohortAnnotationDefinition(enumAnnotationDefinition);
    DbCohortAnnotationEnumValue enumValue3 =
        new DbCohortAnnotationEnumValue()
            .name("a")
            .order(2)
            .cohortAnnotationDefinition(enumAnnotationDefinition);
    enumAnnotationDefinition.getEnumValues().add(enumValue1);
    enumAnnotationDefinition.getEnumValues().add(enumValue2);
    enumAnnotationDefinition.getEnumValues().add(enumValue3);
    enumAnnotationDefinition = cohortAnnotationDefinitionDao.save(enumAnnotationDefinition);

    DbCohortAnnotationDefinition booleanAnnotationDefinition =
        new DbCohortAnnotationDefinition()
            .cohortId(COHORT_ID)
            .columnName("boolean")
            .annotationTypeEnum(AnnotationType.BOOLEAN);
    booleanAnnotationDefinition = cohortAnnotationDefinitionDao.save(booleanAnnotationDefinition);

    pca =
        new DbParticipantCohortAnnotation()
            .cohortAnnotationDefinitionId(
                booleanAnnotationDefinition.getCohortAnnotationDefinitionId())
            .cohortReviewId(cohortReviewId)
            .participantId(participantId)
            .annotationValueBoolean(Boolean.TRUE);
    pca1 =
        new DbParticipantCohortAnnotation()
            .cohortAnnotationDefinitionId(
                enumAnnotationDefinition.getCohortAnnotationDefinitionId())
            .cohortReviewId(cohortReviewId)
            .participantId(participantId)
            .annotationValueEnum("test");
    pca1.setCohortAnnotationEnumValue(enumValue1);
    pca = participantCohortAnnotationDao.save(pca);
    pca1 = participantCohortAnnotationDao.save(pca1);
  }

  @Test
  public void save() {
    assertThat(participantCohortAnnotationDao.findById(pca.getAnnotationId()).get()).isEqualTo(pca);
  }

  @Test
  public void findByCohortReviewIdAndCohortAnnotationDefinitionIdAndParticipantId() {
    assertThat(
            participantCohortAnnotationDao
                .findByCohortReviewIdAndCohortAnnotationDefinitionIdAndParticipantId(
                    cohortReviewId, pca.getCohortAnnotationDefinitionId(), participantId))
        .isEqualTo(pca);
  }

  @Test
  public void findByAnnotationIdAndCohortReviewIdAndParticipantId() {
    assertThat(
            participantCohortAnnotationDao.findByAnnotationIdAndCohortReviewIdAndParticipantId(
                pca.getAnnotationId(), cohortReviewId, participantId))
        .isEqualTo(pca);
  }

  @Test
  public void findByCohortReviewIdAndParticipantId() {
    List<DbParticipantCohortAnnotation> annotations =
        participantCohortAnnotationDao.findByCohortReviewIdAndParticipantId(
            cohortReviewId, participantId);
    assertThat(annotations.size()).isEqualTo(2);
    assertThat(annotations.get(0)).isEqualTo(pca);
    assertThat(annotations.get(1)).isEqualTo(pca1);
    assertThat(
            annotations
                .get(1)
                .getCohortAnnotationEnumValue()
                .getCohortAnnotationDefinition()
                .getEnumValues())
        .isEqualTo(
            pca1.getCohortAnnotationEnumValue().getCohortAnnotationDefinition().getEnumValues());
  }
}

File: api/src/test/java/org/pmiops/workbench/db/dao/ParticipantCohortAnnotationDaoTest.kt
package org.pmiops.workbench.db.dao

import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.pmiops.workbench.db.model.CohortAnnotationDefinition
import org.pmiops.workbench.db.model.CohortAnnotationEnumValue
import org.pmiops.workbench.db.model.ParticipantCohortAnnotation
import org.pmiops.workbench.model.AnnotationType
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.liquibase.LiquibaseAutoConfiguration
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.context.annotation.Import
import org.springframework.test.annotation.DirtiesContext
import org.springframework.test.context.junit4.SpringRunner
import org.springframework.transaction.annotation.Transactional

@RunWith(SpringRunner::class)
@DataJpaTest
@Import(LiquibaseAutoConfiguration::class)
@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
@Transactional
class ParticipantCohortAnnotationDaoTest {
    @Autowired
    private val participantCohortAnnotationDao: ParticipantCohortAnnotationDao? = null
    @Autowired
    private val cohortAnnotationDefinitionDao: CohortAnnotationDefinitionDao? = null
    private var pca: ParticipantCohortAnnotation? = null
    private var pca1: ParticipantCohortAnnotation? = null
    private val cohortReviewId = 3L
    private val participantId = 4L

    @Before
    @Throws(Exception::class)
    fun setUp() {
        val enumAnnotationDefinition = CohortAnnotationDefinition()
                .cohortId(COHORT_ID)
                .columnName("enum")
                .annotationTypeEnum(AnnotationType.ENUM)
        val enumValue1 = CohortAnnotationEnumValue()
                .name("z")
                .order(0)
                .cohortAnnotationDefinition(enumAnnotationDefinition)
        val enumValue2 = CohortAnnotationEnumValue()
                .name("r")
                .order(1)
                .cohortAnnotationDefinition(enumAnnotationDefinition)
        val enumValue3 = CohortAnnotationEnumValue()
                .name("a")
                .order(2)
                .cohortAnnotationDefinition(enumAnnotationDefinition)
        enumAnnotationDefinition.enumValues.add(enumValue1)
        enumAnnotationDefinition.enumValues.add(enumValue2)
        enumAnnotationDefinition.enumValues.add(enumValue3)
        cohortAnnotationDefinitionDao!!.save(enumAnnotationDefinition)

        val booleanAnnotationDefinition = CohortAnnotationDefinition()
                .cohortId(COHORT_ID)
                .columnName("boolean")
                .annotationTypeEnum(AnnotationType.BOOLEAN)

        pca = ParticipantCohortAnnotation()
                .cohortAnnotationDefinitionId(
                        booleanAnnotationDefinition.cohortAnnotationDefinitionId)
                .cohortReviewId(cohortReviewId)
                .participantId(participantId)
                .annotationValueBoolean(java.lang.Boolean.TRUE)
        pca1 = ParticipantCohortAnnotation()
                .cohortAnnotationDefinitionId(
                        enumAnnotationDefinition.cohortAnnotationDefinitionId)
                .cohortReviewId(cohortReviewId)
                .participantId(participantId)
                .annotationValueEnum("test")
        pca1!!.cohortAnnotationEnumValue = enumValue1
        participantCohortAnnotationDao!!.save<ParticipantCohortAnnotation>(pca)
        participantCohortAnnotationDao.save<ParticipantCohortAnnotation>(pca1)
    }

    @Test
    @Throws(Exception::class)
    fun save() {
        assertEquals(pca, participantCohortAnnotationDao!!.findOne(pca!!.annotationId))
    }

    @Test
    @Throws(Exception::class)
    fun findByCohortReviewIdAndCohortAnnotationDefinitionIdAndParticipantId() {
        assertEquals(
                pca,
                participantCohortAnnotationDao!!
                        .findByCohortReviewIdAndCohortAnnotationDefinitionIdAndParticipantId(
                                cohortReviewId, pca!!.cohortAnnotationDefinitionId!!, participantId))
    }

    @Test
    @Throws(Exception::class)
    fun findByAnnotationIdAndCohortReviewIdAndParticipantId() {
        assertEquals(
                pca,
                participantCohortAnnotationDao!!.findByAnnotationIdAndCohortReviewIdAndParticipantId(
                        pca!!.annotationId!!, cohortReviewId, participantId))
    }

    @Test
    @Throws(Exception::class)
    fun findByCohortReviewIdAndParticipantId() {
        val annotations = participantCohortAnnotationDao!!.findByCohortReviewIdAndParticipantId(
                cohortReviewId, participantId)
        assertEquals(2, annotations.size.toLong())
        assertEquals(pca, annotations[0])
        assertEquals(pca1, annotations[1])
        assertEquals(
                pca1!!.cohortAnnotationEnumValue!!.cohortAnnotationDefinition!!.enumValues,
                annotations[1]
                        .cohortAnnotationEnumValue!!
                        .cohortAnnotationDefinition!!
                        .enumValues)
    }

    companion object {

        private val COHORT_ID: Long = 1
    }
}
